std/experimental/allocator/building_blocks/ascending_page_allocator.d

```diff
@@ -0,0 +1,418 @@
+    }
+
+    /**
+    Marks the allocator unavailable for further allocations and sets the ``valid``
+    flag to ``false``, which unmaps the virtual address range when all memory is deallocated.
+    */
+    void invalidate()
+    {
+        valid = false;
+        if (pagesUsed == 0) {
+            version(Posix)
+            {
+                import core.sys.posix.sys.mman : munmap;
+                munmap(data, numPages * pageSize);
+            }
+            else version(Windows)
+            {
+                import core.sys.windows.windows : VirtualFree, MEM_RELEASE;
+                VirtualFree(data, 0, MEM_RELEASE);
+            }
+            data = null;
+        }
+    }
+
+    /**
+    Returns the available size for further allocations in bytes.
+    */
+    size_t getAvailableSize()
+    {
+        return numPages * pageSize + data - offset;
+    }
+
+    /**
+    If the passed buffer is not the last allocation, then ``delta`` can be
+    at most the number of bytes left on the last page.
+    Otherwise, we can expand the last allocation until the end of the virtual
+    address range.
+    */
+    bool expand(ref void[] b, size_t delta)
+    {
+        import std.exception : enforce;
+
+        if (!delta) return true;
+        if (!b.ptr) return false;
+
+        size_t goodSize = goodAllocSize(b.length);
+        size_t bytesLeftOnPage = goodSize - b.length;
+        if (b.ptr + goodSize != offset && delta > bytesLeftOnPage)
+            return false;
+
+        size_t extraPages = 0;
+
+        if (delta > bytesLeftOnPage)
+        {
+            extraPages = goodAllocSize(delta - bytesLeftOnPage) / pageSize;
+        }
+        else
+        {
+            b = cast(void[]) b.ptr[0 .. b.length + delta];
+            return true;
+        }
+
+        if (extraPages > numPages)
+            return false;
+
+        if (offset - data > pageSize * (numPages - extraPages))
+            return false;
+
+        version(Posix)
+        {
+            import core.sys.posix.sys.mman : mprotect, PROT_READ, PROT_WRITE;
+            auto ret = mprotect(offset, extraPages * pageSize, PROT_READ | PROT_WRITE);
+            enforce(ret == 0, "Failed to expand, mprotect failure");
+        }
+        else version(Windows)
+        {
+            import core.sys.windows.windows : VirtualProtect, PAGE_READWRITE;
+            uint oldProtect;
+            auto ret = VirtualProtect(offset, extraPages * pageSize, PAGE_READWRITE, &oldProtect);
+            enforce(ret != 0, "Failed to expand, VirtualProtect failure");
+        }
+
+        pagesUsed += extraPages;
+        offset += extraPages * pageSize;
+        b = cast(void[]) b.ptr[0 .. b.length + delta];
+        return true;
+    }
+
+    /**
+    First it tries to ``expand`` to satisfy ``newSize``.
+    If not possible, it allocates a new buffer of length ``newSize``,
+    copies the contents and deallocates the previous buffer.
+    */
+    bool reallocate(ref void[] b, size_t newSize)
```

Comment:
If this is not different from the stock reallocate, you could remove and just use that.